---
title: 'Lab 15: Transcriptomics'
author: 'Nicholas Do (PID: A15053002)'
date: "11/16/2021"
output: pdf_document
---

# 2. Import countData and colData

```{r}
counts <- read.csv("airway_scaledcounts.csv", row.names=1)
metadata <-  read.csv("airway_metadata.csv")
```

```{r}
head(counts)
head(metadata)
```


> Q1. How many genes are in this dataset?

There are 38694 genes in the dataset.

> Q2. How many ‘control’ cell lines do we have?

```{r}
sum(metadata$dex == "control")
```

There are 4 control cell lines.

# 3. Toy differential gene expression

$DEMO ONLY. NEVER ACTUALLY DO THIS$

```{r}
control <- metadata[metadata[,"dex"]=="control",]
control.counts <- counts[ ,control$id]
control.mean <- rowSums( control.counts )/4 
head(control.mean)
```

> Q3. How would you make the above code in either approach more robust?

Determine the mean using the actual number of control or treated samples or a function
instead of a hard coded number.



> Q4. Follow the same procedure for the treated samples (i.e. calculate the mean per gene across drug treated samples and assign to a labeled vector called treated.mean)


```{r}
#Do this for the treated samples
treated <- metadata[metadata[,"dex"]=="treated",]
treated.counts <- counts[ ,treated$id]
treated.mean <- rowSums( treated.counts )/4 
head(treated.mean)
```

Combine meancount data
```{r}
meancounts <- data.frame(control.mean, treated.mean)
colSums(meancounts)

```

> Q5 (a). Create a scatter plot showing the mean of the treated samples against the mean of the control samples. Your plot should look something like the following.


```{r}
plot(meancounts$control.mean, meancounts$treated.mean)
```

> Q5 (b).You could also use the ggplot2 package to make this figure producing the plot below. What geom_?() function would you use for this plot?

geom_point()


> Q6. Try plotting both axes on a log scale. What is the argument to plot() that allows you to do this?


```{r}
plot(meancounts$control.mean, meancounts$treated.mean, log="xy")

```

```{r}
#calculate log2foldchange
meancounts$log2fc <- log2(meancounts[,"treated.mean"]/meancounts[,"control.mean"])
head(meancounts)
```

Filter results:
```{r}
zero.vals <- which(meancounts[,1:2]==0, arr.ind=TRUE)

head(zero.vals)
to.rm <- unique(zero.vals[,1])
head(to.rm)
mycounts <- meancounts[-to.rm,]
head(mycounts)
```


> Q7. What is the purpose of the arr.ind argument in the which() function call above? Why would we then take the first column of the output and need to call the unique() function?


The arr.ind argument allows the which() function to return the array indices, which we
will work with in order to filter our results. We call the unique() function using the
first column of the output in order to remove any duplicate rows.


Filter for regulation:
```{r}
up.ind <- mycounts$log2fc > 2
down.ind <- mycounts$log2fc < (-2)
```

> Q8. Using the up.ind vector above can you determine how many up regulated genes we have at the greater than 2 fc level?

```{r}
#count(up.ind == "TRUE")
```
We have 250 up-regulated genes greater than 2 fc level.

> Q9. Using the down.ind vector above can you determine how many down regulated genes we have at the greater than 2 fc level? 

```{r}
#count(down.ind == "TRUE")
```
We have 367 down-regulated genes greater than 2 fc level.

> Q10. Do you trust these results? Why or why not?

I would not yet trust these results because we have not yet determined
the statistical significance of our values, and there may be a large 
amount of variance involved. We need to analyze using DESeq2.

# 4. DESeq2 Analysis

```{r}
library(DESeq2)
citation("DESeq2")
```


Build dataset object
```{r}
dds <- DESeqDataSetFromMatrix(countData=counts, 
                              colData=metadata, 
                              design=~dex)
dds
```

```{r}
#Nothing exists yet, need to run DESeq()
#results(dds)
```


```{r}
dds <- DESeq(dds)
```

Look at the results!
```{r}
res <- results(dds)
res
```


Convert to dataframe:
```{r}
resdf <- as.data.frame(res)
View(resdf)
```

summarize:
```{r}
summary(res)
```

```{r}
res05 <- results(dds, alpha=0.05)
summary(res05)
```

# 5. Adding annotation data

```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")

columns(org.Hs.eg.db)
```

Annotate using mapIds():
```{r}
res$symbol <- mapIds(org.Hs.eg.db,
                     keys=row.names(res), # Our genenames
                     keytype="ENSEMBL",        # The format of our genenames
                     column="SYMBOL",          # The new format we want to add
                     multiVals="first")

head(res)

```

> Q11. Run the mapIds() function two more times to add the Entrez ID and UniProt accession and GENENAME as new columns called res$entrez, res$uniprot and res$genename.

```{r}
res$entrez <- mapIds(org.Hs.eg.db,
                     keys=row.names(res),
                     column="ENTREZID",
                     keytype="ENSEMBL",
                     multiVals="first")

res$uniprot <- mapIds(org.Hs.eg.db,
                     keys=row.names(res),
                     column="UNIPROT",
                     keytype="ENSEMBL",
                     multiVals="first")

res$genename <- mapIds(org.Hs.eg.db,
                     keys=row.names(res),
                     column="GENENAME",
                     keytype="ENSEMBL",
                     multiVals="first")

head(res)
```


Arrange and view by adjPval
```{r}
ord <- order( res$padj )
#View(res[ord,])
head(res[ord,])
```

```{r}
write.csv(res[ord,], "deseq_results.csv")
```


# 6. Data Visualization 


Dull Volcano Plot
```{r}
plot( res$log2FoldChange,  -log(res$padj), 
      xlab="Log2(FoldChange)",
      ylab="-Log(P-value)")
```


More useful plot:
```{r}
plot( res$log2FoldChange,  -log(res$padj), 
 ylab="-Log(P-value)", xlab="Log2(FoldChange)")

# Add some cut-off lines
abline(v=c(-2,2), col="darkgray", lty=2)
abline(h=-log(0.05), col="darkgray", lty=2)
```

Color the points:
```{r}
# Setup our custom point color vector 
mycols <- rep("gray", nrow(res))
mycols[ abs(res$log2FoldChange) > 2 ]  <- "red" 

inds <- (res$padj < 0.01) & (abs(res$log2FoldChange) > 2 )
mycols[ inds ] <- "blue"

# Volcano plot with custom colors 
plot( res$log2FoldChange,  -log(res$padj), 
 col=mycols, ylab="-Log(P-value)", xlab="Log2(FoldChange)" )

# Cut-off lines
abline(v=c(-2,2), col="gray", lty=2)
abline(h=-log(0.1), col="gray", lty=2)
```

Enhanced volcano
```{r}
library(EnhancedVolcano)
x <- as.data.frame(res)

EnhancedVolcano(x,
    lab = x$symbol,
    x = 'log2FoldChange',
    y = 'pvalue')
```


# 7. Pathway analysis

Set up KEGG Datasets
```{r}
library(pathview)
library(gage)
library(gageData)

data(kegg.sets.hs)

# Examine the first 2 pathways in this kegg set for humans
head(kegg.sets.hs, 2)
```


```{r}
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges)
```


```{r}
# Get the results
keggres = gage(foldchanges, gsets=kegg.sets.hs)

attributes(keggres)
```

```{r}
# Look at the first three down (less) pathways
head(keggres$less, 3)
```


```{r}
pathview(gene.data=foldchanges, pathway.id="hsa05310")
```

![](hsa05310.pathview.png)

> Q12. Can you do the same procedure as above to plot the pathview figures for the top 2 down-reguled pathways?

Yes, you can.

Graff-versus-host disease
```{r}
pathview(gene.data=foldchanges, pathway.id="hsa05332")
```

![](hsa05332.pathview.png)

Type 1 diabetes
```{r}
pathview(gene.data=foldchanges, pathway.id="hsa04940")
```

![](hsa04940.pathview.png)










